src/librustc_mir/const_eval.rs

```diff
@@ -336,6 +336,20 @@ impl<'mir, 'tcx> interpret::Machine<'mir, 'tcx> for CompileTimeInterpreter<'mir,
         _unwind: Option<mir::BasicBlock> // unwinding is not supported in consts
     ) -> InterpResult<'tcx, Option<&'mir mir::Body<'tcx>>> {
         debug!("eval_fn_call: {:?}", instance);
+
+        // If this function is a `const fn` then as an optimization we can query this
+        // evaluation immediately.
+        //
+        // For the moment we only do this for functions which take no arguments
+        // (or all arguments are ZSTs) so that we don't memoize too much.
+        if ecx.tcx.is_const_fn_raw(instance.def.def_id()) &&
+            args.iter().all(|a| a.layout.is_zst())
+        {
+            let gid = GlobalId { instance, promoted: None };
+            ecx.eval_const_fn_call(gid, ret)?;
+            return Ok(None);
+        }
+
         // Only check non-glue functions
         if let ty::InstanceDef::Item(def_id) = instance.def {
             // Execution might have wandered off into other crates, so we cannot do a stability-
```

Comment:
There's a `is_const_fn_raw` check below, maybe you can move your check down into there?